### PR TITLE
ARROW-6945: [Rust][Integration] Run rust integration tests

### PR DESCRIPTION
**Pull request by nealrichardson**
FYI @andygrove. I looked at the integration test output and saw that Rust tests weren't being run at all, so I'm trying to enable them.

### CODE REVIEW

**Comment by andygrove** [comment]
Thanks @nealrichardson .. so far I've just been running them manually with (from memory) `archery integration --with-java=true --with-rust=true`

**Comment by nealrichardson** [comment]
Ok @andygrove I got them running (in a sense) but they all fail. See https://github.com/apache/arrow/pull/7297/checks?check_run_id=717518551

Error message is all `FileNotFoundError: [Errno 2] No such file or directory: '/arrow/rust/target/debug/arrow-json-integration-test'` so something isn't getting set up correctly. I'm hoping you know what that is. Can you take over from here?

**Comment by andygrove** [comment]
Thanks @nealrichardson .. `cargo build` looks like it ran ok, so I don't know why the binary doesn't exist in that location (which looks correct). I will take a look at this tonight and see what the issue is.

cc @vertexclique Since you were interested in helping with this.

**Comment by nealrichardson** [comment]
Side note: does this need a new JIRA or does one already exist. Maybe use https://issues.apache.org/jira/browse/ARROW-6945 for this?

**Comment by nealrichardson** [comment]
Did you say that integration tests were a separate crate now? Maybe that wasn't built. I just added `ci/scripts/rust_build.sh`

**Comment by andygrove** [comment]
Yes it is a new crate but part of the workspace so the top level `cargo build` should build it.

**Comment by andygrove** [comment]
I used docker compose to run this locally and could not reproduce the issue. The tests ran for me.

Brief sample output:

```
["/arrow/rust/target/debug/arrow-json-integration-test", "--integration", "--arrow=/tmp/tmpuv1jy3t6/8766f48c_generated_nested_dictionary.json_as_file", "--json=/tmp/arrow-integration-ewn5k3l5/generated_nested_dictionary.json", "--mode=JSON_TO_ARROW"]
Converting /tmp/arrow-integration-ewn5k3l5/generated_nested_dictionary.json to /tmp/tmpuv1jy3t6/8766f48c_generated_nested_dictionary.json_as_file
Error: JsonError("data type Dictionary(Int8, List(Dictionary(Int8, Utf8))) not supported")
```

**Comment by andygrove** [comment]
I'm running this again, redirecting output to a log file. I will find a way to share this tomorrow.

**Comment by andygrove** [comment]
Full logs are attached to the JIRA

**Comment by nevi-me** [comment]
I have made progress on large arrays (https://github.com/apache/arrow/compare/master...nevi-me:rust-large-lists), the code is very repetitive so I still need to clean it up.

When large lists are implemented, there are 2 issues that become visible:

- We don't handle empty batch tests correctly (we fail instead of returning something)
- There's a bug with how we write `NullArray` which results in any arrays that are written after, not being read by the consumer.

I think outside of the above, we can disable other tests.

**Comment by nealrichardson** [comment]
@nevi-me @andygrove is it expected that everything except `primitive_no_batches` fails for Rust? If so, we can skip everything else, and y'all can unskip them iteratively when you get to them. If that's not expected, you should probably debug that before merging this.

**Comment by andygrove** [comment]
@nealrichardson I think that would be a reasonable assumption to make at this point. I'd be happy with this merged even if only a small number of tests are passing and we can iterate from there, Sorry, I'd help more but am working long hours in the day job currently. I should have time before the release to re-engage with this.

**Comment by andygrove** [comment]
@nevi-me @nealrichardson I have some time available this week. How can I help with this?

**Comment by nevi-me** [comment]
@andygrove thanks. I pushed done if what I worked on over the weekend. The main problem seems to be that we don't read all record batches from the Arrow files. As a result we end up getting "no more record batches" errors.

I haven't yet been able to figure out why we don't read all batches, the block sizes indicate the correct batch counts, but the reader just fails to read them all 😞

**Comment by nevi-me** [comment]
Oh, there's also a TODO on array data comparisons. I think we only compare the array lengths and types for now; but not their data.

File: dev/archery/archery/integration/datagen.py [hunk 1]
```diff
@@ -1492,21 +1492,25 @@ def _temp_path():
 
         generate_primitive_large_offsets_case([17, 20])
         .skip_category('Go')
-        .skip_category('JS'),
+        .skip_category('JS')
+        .skip_category('Rust'),
 
         generate_null_case([10, 0])
+        .skip_category('Rust')
```
Comment (nealrichardson) [inline dev/archery/archery/integration/datagen.py:1499]: Didn't I see a patch merge recently that added null support? If you rebase, can you remove this skip?

**Comment by nealrichardson** [comment]
Ok there was one other test needing to be skipped, which I've done, and now the tests "pass". Should we merge this and progressively unskip tests as you can?

**Comment by andygrove** [comment]
Yes, that would be great. Thanks!

**Comment by nevi-me** [comment]
> Ok there was one other test needing to be skipped, which I've done, and now the tests "pass". Should we merge this and progressively unskip tests as you can?

Yes please

**Review by nevi-me — APPROVED** [review]

**Review by andygrove — APPROVED** [review]

**Pull request closed without merging**